packages/plugin-ext/src/main/browser/webview/webview.ts

```diff
@@ -13,316 +13,571 @@
+export const WebviewWidgetExternalEndpoint = Symbol('WebviewWidgetExternalEndpoint');
+
+@injectable()
+export class WebviewWidget extends BaseWidget implements StatefulWidget {
+
+    private static readonly standardSupportedLinkSchemes = new Set([
+        Schemes.HTTP,
+        Schemes.HTTPS,
+        Schemes.MAILTO,
+        Schemes.VSCODE
+    ]);
+
+    static FACTORY_ID = 'plugin-webview';
+
+    protected element: HTMLIFrameElement | undefined;
+
     // tslint:disable-next-line:max-line-length
-    // XXX This is a hack to be able to tack the mouse events when drag and dropping the widgets. On `mousedown` we put a transparent div over the `iframe` to avoid losing the mouse tacking.
-    protected readonly transparentOverlay: HTMLElement;
-
-    constructor(title: string,
-        private options: WebviewWidgetOptions,
-        private eventDelegate: WebviewEvents,
-        protected readonly mouseTracker: ApplicationShellMouseTracker) {
-        super();
+    // XXX This is a hack to be able to tack the mouse events when drag and dropping the widgets.
+    // On `mousedown` we put a transparent div over the `iframe` to avoid losing the mouse tacking.
+    protected transparentOverlay: HTMLElement;
+
+    @inject(WebviewWidgetIdentifier)
+    readonly identifier: WebviewWidgetIdentifier;
+
+    @inject(WebviewWidgetExternalEndpoint)
+    readonly externalEndpoint: string;
+
+    @inject(ApplicationShellMouseTracker)
+    protected readonly mouseTracker: ApplicationShellMouseTracker;
+
+    @inject(WebviewEnvironment)
+    protected readonly environment: WebviewEnvironment;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    @inject(KeybindingRegistry)
+    protected readonly keybindings: KeybindingRegistry;
+
+    @inject(PluginSharedStyle)
+    protected readonly sharedStyle: PluginSharedStyle;
+
+    @inject(WebviewThemeDataProvider)
+    protected readonly themeDataProvider: WebviewThemeDataProvider;
+
+    @inject(ExternalUriService)
+    protected readonly externalUriService: ExternalUriService;
+
+    @inject(OutputChannelManager)
+    protected readonly outputManager: OutputChannelManager;
+
+    @inject(WebviewPreferences)
+    protected readonly preferences: WebviewPreferences;
+
+    @inject(WebviewResourceLoader)
+    protected readonly resourceLoader: WebviewResourceLoader;
+
+    @inject(WebviewResourceCache)
+    protected readonly resourceCache: WebviewResourceCache;
+
+    viewState: WebviewPanelViewState = {
+        visible: false,
+        active: false,
+        position: 0
+    };
+
+    protected html = '';
+
+    protected _contentOptions: WebviewContentOptions = {};
+    get contentOptions(): WebviewContentOptions {
+        return this._contentOptions;
+    }
+
+    protected _state: string | undefined;
+    get state(): string | undefined {
+        return this._state;
+    }
+
+    viewType: string;
+    options: WebviewPanelOptions = {};
```

Comment:
looks like we don't support `options.customClasses` or `options.tryRestoreScrollPosition`. do we have a GH issue for it ?

From vs code api :
```
export interface WebviewOptions {
	readonly customClasses?: string;
	readonly enableFindWidget?: boolean;
	readonly tryRestoreScrollPosition?: boolean;
	readonly retainContextWhenHidden?: boolean;
}
```

Reply:
Could you find that they are public exposed on vscode.d.ts? They seem be to be for internal views, which are not of interest for us.

Reply:
ah you are right. this interface is not exposed.